### PR TITLE
Reduce the number of cache reads and writes in OpsCache

### PR DESCRIPTION
**Pull request by jatgarg**
## Description

ADO task link: https://dev.azure.com/fluidframework/internal/_workitems/edit/5556

Reduce cache reads by bailing out early if we don't get any ops from the current batch.

Reduce the number of cache write by not flushing out an empty batch or a batch which has empty slots at both start and end as this will mostly not satisfy the need during read.

### CODE REVIEW

File: packages/drivers/odsp-driver/src/opsCaching.ts [hunk 1]
```diff
@@ -65,7 +65,14 @@ export class OpsCache {
...
 	public flushOps() {
 		for (const [key, value] of this.batches) {
-			if (value === null || !value.dirty) {
+			// Don't flush if the batch has no ops, already flushed or has empty slots at both beginning and end.
+			if (
+				value === null ||
+				!value.dirty ||
+				value.batchData.length === 0 ||
+				(value.batchData[0] === undefined &&
+					value.batchData[value.batchData.length - 1] === undefined)
```
Comment (rajatch-ff) [inline packages/drivers/odsp-driver/src/opsCaching.ts:74]: Do we have tests for each of these condition to test each codepath? It's probably easy to just have different types of mockData to execute each condition.

Comment (jatgarg) [reply packages/drivers/odsp-driver/src/opsCaching.ts:74]: Yes we already have tests related to ops caching but 1 of them failed which I expected, fixed that and added couple of more tests.

Comment (NicholasCouri) [reply packages/drivers/odsp-driver/src/opsCaching.ts:74]: Jatin, just so I can understand it, what kind of scenario would match this condition: 
(value.batchData[0] === undefined &&
					value.batchData[value.batchData.length - 1] === undefined  ?

Comment (jatgarg) [reply packages/drivers/odsp-driver/src/opsCaching.ts:74]: So we push ops in opsCache based on timer too. So lets say we have batches of size 100 and we started getting ops from 30 (for ex lets say we loaded from seq number 29), then we timer out and we only got ops from 30-80 by then, in that case we would not want to flush those ops as the batch is not filled yet.

**Review by NicholasCouri — APPROVED** [review]

**Pull request merged**